File: 07threads/somma_primi_files.c
/*
 * Esempio paradigma produttori consumatori
 * con produttori e consumatori multipli
 * 
 * I produttori leggono interi dai file e i consumatori
 * cacolano il numero e la somma dei primi
 * 
 * il numero di produttori è uguale al numero di file di input
 * il numero di consumatori è stabilito all'inizo del main() 
 *  
 * Usare il numeri.py per generare lunghi elenchi di interi positivi su cui testare il programma
 *  * 
 * */
#include "xerrori.h"

#define Buf_size 10

// funzione per stabilire se n e' primo  
bool primo(int n)
{
  if(n<2) return false;
  if(n%2==0) return (n==2);
  for (int i=3; i*i<=n; i += 2)
      if(n%i==0) return false;
  return true;
}

// struct contenente i parametri dei thread consumatori
typedef struct {
  int quanti;    // parametro di output
  long somma;    // parametro di output
  int *buffer;   // variabili per schema prod/consumatori
  int *pcindex;  // puntatore a cindex 
  pthread_mutex_t *pmutex;
  sem_t *sem_free_slots;
  sem_t *sem_data_items;  
} datic;


// struct contenente i parametri dei thread produttori
typedef struct {
  char *nomefile;  // nome file da gestire 
  int *buffer;     // variabili per schema prod/consumatori
  int *ppindex;    // puntatore a pindex
  pthread_mutex_t *pmutex;
  sem_t *sem_free_slots;
  sem_t *sem_data_items;  
} datip;



// funzione eseguita dai thread consumatori
void *cbody(void *arg)
{  
  datic *a = (datic *)arg; 
  a->quanti = 0;
  a->somma = 0;
  fprintf(stderr,"[C] consumatore partito\n");
  int n;
  do {
    xsem_wait(a->sem_data_items,__LINE__,__FILE__);
    xpthread_mutex_lock(a->pmutex,__LINE__,__FILE__);
    n = a->buffer[*(a->pcindex) % Buf_size];
    *(a->pcindex) +=1;
    xpthread_mutex_unlock(a->pmutex,__LINE__,__FILE__);
    xsem_post(a->sem_free_slots,__LINE__,__FILE__);
    if(n>0 && primo(n)) {
      a->quanti++;
      a->somma += n;
    }
  } while(n!= -1);
  pthread_exit(NULL); 
}     


// funzione eseguita dai thread produttori
void *pbody(void *arg)
{  
  datip *a = (datip *)arg; 
  fprintf(stderr,"[P] produttore con file %s partito\n",a->nomefile);
  // apro il file che ho ricevuto da main
  FILE *f = fopen(a->nomefile,"rt");
  if(f==NULL) pthread_exit(NULL); 
  int e,n;
  while(true) {
    e = fscanf(f,"%d", &n);
    if(e!=1) break; // se il valore e' letto correttamente e==1
    assert(n>0);    // i valori del file devono essere positivi
    xsem_wait(a->sem_free_slots,__LINE__,__FILE__);
    xpthread_mutex_lock(a->pmutex,__LINE__,__FILE__);
    a->buffer[(*a->ppindex)++ % Buf_size]= n;
    xpthread_mutex_unlock(a->pmutex,__LINE__,__FILE__);
    xsem_post(a->sem_data_items,__LINE__,__FILE__);
  }
  fclose(f);
  pthread_exit(NULL); 
}     




int main(int argc, char *argv[])
{
  // leggi input
  if(argc<2) {
    printf("Uso\n\t%s file1 [file2 file3 ... filek]\n", argv[0]);
    exit(1);
  }
  int cons = 3;  // numero thread consumatori
  assert(cons>0);
  int prod = argc-1;  // numero thread produttori
  assert(prod>0);
  
  int tot_primi = 0;
  long tot_somma = 0;
  // threads related
  int buffer[Buf_size];
  int cindex=0, pindex=0;
  // mutex x consumatori
  pthread_mutex_t cmu = PTHREAD_MUTEX_INITIALIZER;
  // mutex x produttori
  pthread_mutex_t pmu = PTHREAD_MUTEX_INITIALIZER;
  // semafori x produttori e consumatori 
  sem_t sem_free_slots, sem_data_items;
  xsem_init(&sem_free_slots,0,Buf_size,__LINE__,__FILE__);
  xsem_init(&sem_data_items,0,0,__LINE__,__FILE__);
  pthread_t tc[cons];
  datic ac[cons];
  for(int i=0;i<cons;i++) {
    // faccio partire il thread i
    ac[i].buffer = buffer;
    ac[i].pcindex = &cindex;
    ac[i].pmutex = &cmu;
    ac[i].sem_data_items = &sem_data_items;
    ac[i].sem_free_slots = &sem_free_slots;
    xpthread_create(&tc[i],NULL,cbody,&ac[i],__LINE__,__FILE__);
  }
  fprintf(stderr,"Thread consumatori creati");
  pthread_t tp[prod];
  datip ap[prod];
  for(int i=0;i<prod;i++) {
    // faccio partire il thread produttore i
    ap[i].buffer = buffer;
    ap[i].ppindex = &pindex;
    ap[i].pmutex = &pmu;
    ap[i].sem_data_items = &sem_data_items;
    ap[i].sem_free_slots = &sem_free_slots;
    ap[i].nomefile = argv[i+1];
    xpthread_create(&tp[i],NULL,pbody,&ap[i],__LINE__,__FILE__);
  }
  fprintf(stderr,"Thread produttori creati");
  // join dei thread produttori
  for(int i=0;i<prod;i++) {
    xpthread_join(tp[i],NULL,__LINE__,__FILE__);
  }
  // terminazione thread consumatori
  for(int i=0;i<cons;i++) {
    xsem_wait(&sem_free_slots,__LINE__,__FILE__);
    buffer[pindex++ % Buf_size]= -1;
    xsem_post(&sem_data_items,__LINE__,__FILE__);
  }
  puts("Valori di terminazione scritti");
  // join dei thread consumatori e calcolo risulato
  for(int i=0;i<cons;i++) {
    xpthread_join(tc[i],NULL,__LINE__,__FILE__);
    tot_primi += ac[i].quanti;
    tot_somma += ac[i].somma;
  }
  xpthread_mutex_destroy(&pmu,__LINE__,__FILE__);
  xpthread_mutex_destroy(&cmu,__LINE__,__FILE__);
  xsem_destroy(&sem_data_items,__LINE__,__FILE__);
  xsem_destroy(&sem_free_slots,__LINE__,__FILE__);
  fprintf(stderr,"Trovati %d primi con somma %ld\n",tot_primi,tot_somma);
  return 0;
}

File: 11pythreads/somma_primi_files.py
#!/usr/bin/env python3
import threading
import queue
import sys
import math

# Funzione per verificare se un numero è primo
def primo(n):
    if n < 2:
        return False
    if n % 2 == 0:
        return n == 2
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False
    return True

# Funzione eseguita dai thread consumatori
def cbody(buffer, buffer_lock, sem_free_slots, sem_data_items, results, thread_id):
    quanti = 0
    somma = 0
    print(f"[C] Consumatore {thread_id} partito")
    while True:
        sem_data_items.acquire()
        with buffer_lock:
            n = buffer.get()
        sem_free_slots.release()
        if n == -1:  # Segnale di terminazione
            break
        if primo(n):
            quanti += 1
            somma += n
    results[thread_id] = (quanti, somma)
    print(f"[C] Consumatore {thread_id} terminato")

# Funzione eseguita dai thread produttori
def pbody(nomefile, buffer, buffer_lock, sem_free_slots, sem_data_items):
    print(f"[P] Produttore con file {nomefile} partito")
    try:
        with open(nomefile, "r") as f:
            for line in f:
                n = int(line.strip())
                assert n > 0  # I valori del file devono essere positivi
                sem_free_slots.acquire()
                with buffer_lock:
                    buffer.put(n)
                sem_data_items.release()
    except FileNotFoundError:
        print(f"Errore: file {nomefile} non trovato")
    print(f"[P] Produttore con file {nomefile} terminato")

def main():
    if len(sys.argv) < 2:
        print(f"Uso: {sys.argv[0]} file1 [file2 file3 ... filek]")
        sys.exit(1)

    cons = 3  # Numero di thread consumatori
    assert cons > 0
    prod = len(sys.argv) - 1  # Numero di thread produttori
    assert prod > 0

    tot_primi = 0
    tot_somma = 0

    # Buffer condiviso
    buffer = queue.Queue(maxsize=10)
    buffer_lock = threading.Lock()

    # Semafori
    sem_free_slots = threading.Semaphore(10)
    sem_data_items = threading.Semaphore(0)

    # Creazione e avvio dei thread consumatori
    results = [None] * cons
    consumer_threads = []
    for i in range(cons):
        t = threading.Thread(target=cbody, args=(buffer, buffer_lock, sem_free_slots, sem_data_items, results, i))
        consumer_threads.append(t)
        t.start()
    print("Thread consumatori creati")

    # Creazione e avvio dei thread produttori
    producer_threads = []
    for i in range(prod):
        t = threading.Thread(target=pbody, args=(sys.argv[i + 1], buffer, buffer_lock, sem_free_slots, sem_data_items))
        producer_threads.append(t)
        t.start()
    print("Thread produttori creati")

    # Attesa della terminazione dei thread produttori
    for t in producer_threads:
        t.join()

    # Scrittura dei segnali di terminazione per i consumatori
    for _ in range(cons):
        sem_free_slots.acquire()
        with buffer_lock:
            buffer.put(-1)
        sem_data_items.release()
    print("Valori di terminazione scritti")

    # Attesa della terminazione dei thread consumatori e calcolo del risultato
    for t in consumer_threads:
        t.join()

    for quanti, somma in results:
        tot_primi += quanti
        tot_somma += somma

    print(f"Trovati {tot_primi} primi con somma {tot_somma}")

if __name__ == "__main__":
    main()